determine router expose ports by parsing HTTP_EXPOSE and TCP_EXPOSE per site container

tannerjfco opened the pull request:
## The Problem:
We want to provide support for users to define their own additional services to run as part of a ddev app. We are providing support for multiple extending/overriding compose files in #198, but we also need a way to expose new ports on the router, so that users can define a new service to be available at `sitename.ddev.local:portNo`.

## The Fix:
This PR changes how ports are exposed on the router container. Previously, each port we needed to expose was explicitly defined as a value in the router template. The new approach works by:
- Leveraging the existing `VIRTUAL_PORT` env var used to determine which ports are configured as upstreams in the router.
- Retrieving values of `VIRTUAL_PORT` from all site containers
- Retrieving existing port mappings from the router if it is running
- Reconciling the two, adding any new port mappings from the site containers that aren't present in any current router port mappings

The `VIRTUAL_PORT` env var can accept multiple port definitions for a container via comma separated values. Ports can also be defined in the docker `hostPort:containerPort` format, allowing for the exposed port to differ from the internal container port.

## The Test:
drud/nginx-proxy#10 is required to test this change. This PR is configured to use the image from this PR, so grabbing the binary for this PR is sufficient to test.

You should be able to test this PR in the following ways:
- Test the binary built from this PR  against a site with default ddev configuration. The site should work as normally expected. With this change, no ports are exposed on the router unless a container tells it to, so normal operation should indicate this is working.
- To test `hostPort:containerPort` port definition, you can change the `VIRTUAL_PORT` env var on the web container for a site, changing the mailhog port definition from `8025` to `8050:8025`.
- The sample solr config file at https://github.com/drud/ddev/issues/121#issuecomment-297839323 can be used to expose another service for a site. You should be able to access the solr container at `sitename.ddev.local:8983`

## Automation Overview:
<!-- Please provide an overview of tests introduced by this PR, or an explanation for why no tests are needed. -->
Test cases have been introduced for the new router functionality

## Related Issue Link(s):
#121 ability to use custom containers
drud/nginx-proxy#10 directly relates to this PR - it will need to be merged and a new release tagged prior to this PR coming in.

## Release/Deployment notes:
<!-- Does this affect anything else, or are there ramifications for other code? Does anything have to be done on deployment? -->

beeradb commented on pkg/util/dockerutils.go:
```diff
@@ -219,3 +219,30 @@ func GetContainerHealth(container docker.APIContainers) string {
+	client := GetDockerClient()
+	inspect, err := client.InspectContainer(container.ID)
+	if err != nil {
+		log.Debug("failed to retreive container ", ContainerName(container))
```
Seems like we want to return an error as well?

rfay replied:
s/retreive/retrieve/g

beeradb commented on pkg/util/dockerutils.go:
```diff
@@ -219,3 +219,30 @@ func GetContainerHealth(container docker.APIContainers) string {
+	label := map[string]string{"com.ddev.site-name": sitename}
+	sites, err := FindContainersByLabels(label)
+	if err != nil {
+		log.Fatal("failed to retrieve containers for ", sitename, err)
```
It seems like this should probably return an error instead of a fatal? It's only called by Start(), which itself returns an error. Feels like just bubbling this up to the caller is a better approach

beeradb commented on pkg/util/dockerutils.go:
```diff
@@ -219,3 +219,30 @@ func GetContainerHealth(container docker.APIContainers) string {
+			return strings.TrimPrefix(env, key+"=")
+		}
+	}
+	return ""
```
I think there's a good case for not found being an error, but you have more context than I do on this one

tannerjfco replied:
The intent here was to provide similar usage to what `os.Getenv()` provides. If the key does not exist, it returns an empty string.

beeradb commented on pkg/plugins/platform/templates.go:
```diff
@@ -69,10 +69,8 @@ services:
-      - {{ .mailhogport }}:{{ .mailhogport }}
-      - {{ .dbaport }}:{{ .dbaport }}
-      - {{ .dbport }}:{{ .dbport }}
+      {{ range $port := .ports }}- "{{ $port }}:{{ $port }}"
```
So this is actually a mistake that I introduced without realizing it, and have known would need to be fixed when we get here 😭. The format of the ports directive is "host:container". We only want the host port to change, not the internal container ports. The ports variable probably needs to be a map which maps external -> internal ports.

tannerjfco replied:
Are you sure about that? It seems to me that we would want a 1:1 relation here. If a user defined `VIRTUAL_PORT=8765`, the router would create an upstream listening on port 8765 internally, and we would then want to the router to accept external traffic from 8765.

beeradb replied:
Sometimes this is true, but I'm not sure it holds in general. For user-defined services it makes sense that there would be a 1:1 ratio, but for things like MariaDB/nginx do we really want these changed? If someone wants to change from port 80 on HTTP to port 25049, do we really want to remap that within the pod as well? I'm not sold on that being a great idea.

beeradb commented on pkg/plugins/platform/local.go:
```diff
@@ -310,14 +310,33 @@ func (l *LocalApp) Start() error {
+
+	// Retrieve ports from containers to expose in router
+	var exposePorts []string
+	siteContainers, err := util.GetAppContainers(l.GetName())
```
👏

beeradb commented on pkg/plugins/platform/local_test.go:
```diff
@@ -131,6 +132,14 @@ func TestGetApps(t *testing.T) {
 	}
 }
 
+func TestGetCurrentRouterPorts(t *testing.T) {
```
<pedantic> should have a comment </pedantic>

beeradb commented on pkg/util/dockerutils_test.go:
```diff
@@ -84,3 +85,24 @@ func TestComposeCmd(t *testing.T) {
+	assert.NoError(err)
+
+	env := GetContainerEnv("HOTDOG", container)
+	assert.Equal("superior-to-corndog", env)
```
Fatal error.

rfay commented:
I just now realized how this related to the security issues and desire to use a non-root user to run docker containers.

AFAIK, the only thing that forces us to use root for *anything* in the nginx container is the use of port 80 for nginx. If that were above 1024, nginx could run (and everything else run) as an unprivileged user, and we'd have no reason to be using root in the container. Worth thinking about, and contradicts my assertion yesterday that we should be using standard ports.

tannerjfco commented:
I've updated this PR and introduced drud/nginx-proxy#9 to facilitate defining ports as hostPort:containerPort. This will allow us and developers to define ports as e.g. 80:8080, allowing port 8080 on a container to be exposed as port 80.

rickmanelius commented:
Given the @beeradb suggested a discussion to get consensus, is there a scheduled meeting to have said discussion in order to keep forward progress on this issue?

rfay commented:
I took the liberty of adding the requirement of https://github.com/drud/nginx-proxy/pull/9 to the OP

rfay commented on Makefile:
```diff
@@ -24,7 +24,7 @@ WebTag ?= v0.4.0
 DBTag ?= v0.3.0
 RouterImage ?= drud/nginx-proxy
-RouterTag ?= v0.3.0
+RouterTag ?= router-expose
```
This has to be updated before pull. I'd have been more comfortable with the process if this were either set to the expected tag (v0.4.0?) and a provisional v0.4.0 pushed, or else left alone and the instructions telling us to build ddev with `make RouterTag=router-expose`. But every single way is a bit of a cheat to get around all the moving components, so yours is just as good I'd say. But has to be updated.

rfay commented on pkg/util/util_test.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+var (
+	TestArchiveURL        = "https://github.com/drud/wordpress/archive/v0.4.0.tar.gz"
```
If these are going to be exported, they need a good comment with usage information.

rfay commented on pkg/version/version.go:
```diff
@@ -30,7 +30,7 @@ var DBATag = "v0.2.0"
 
 // RouterTag defines the tag used for the router.
-var RouterTag = "v0.3.0" // Note that this is overridden by make
+var RouterTag = "router-expose" // Note that this is overridden by make
```
Update to real tag before pull.

beeradb commented on pkg/plugins/platform/router.go:
```diff
@@ -84,3 +86,45 @@ func StartDockerRouter() {
+
+// SetRouterPorts determines the router port configuration based on port definitions from
+// site containers and any current configuration from a running router.
+func SetRouterPorts(ports []string) []string {
```
If I understand this correctly it just appends to the current list? What happens if we want to change existing ports, such as port 80/3306. Would this gracefully handle that or maintain the old ports as well?

tannerjfco replied:
Currently this would maintain the existing ports. The intent is to keep port definitions from other sites in place without having to re-query every container for every running site. This likely falls apart if an existing external port gets remapped to a different internal port though.

tannerjfco commented:
OP updated as requested with updated change summary and test instructions

beeradb commented:
@rfay at this point the router is re-templated on every site start, so custom configuration is not possible. I'm not sure we can do this in a sane way and allow for custom router configuration unless we do something like check a separate location for a user defined router manifest and start that one if it exists. I'm not sure that would be of practical value, though.

tannerjfco commented:
I've reworked this to determine the router port mappings from running site containers. This should help ensure ports no longer in use are removed, and we don't end up with an unwieldy list of port mappings on the router container.

The only situation not explicitly addressed here, is if you try to remap an existing external port to a different internal port. In this situation, docker-compose produces a failure with a message indicating the external port is already allocated. I believe this should be sufficient handling for this scenario, as I expect it to be more of an edge case, and the docker-compose error message reasonably describes the issue encountered.

rfay commented:
I wonder if we could add documentation into the docker-compose.yaml template, so people would just innately understand how to extend it based on the in-place docs?

rfay commented:
Tested this with an extra and separate web container exposed on 8089 and it worked, thanks! Note that lots of other things worked too, including ddev ssh -s junk, etc.

```
  junk:
    container_name: local-${DDEV_SITENAME}-junk
    image: $DDEV_WEBIMAGE
    restart: always
    ports:
      - "80"
    working_dir: "/var/www/html/docroot"
    environment:
      - DEPLOY_NAME=local
      - VIRTUAL_HOST=$DDEV_HOSTNAME
      - VIRTUAL_PORT=8089:80
    labels:
      com.ddev.site-name: ${DDEV_SITENAME}
      com.ddev.platform: local
      com.ddev.app-type: drupal7
      com.ddev.docroot: $DDEV_DOCROOT
      com.ddev.approot: $DDEV_APPROOT
      com.ddev.app-url: $DDEV_URL
```

rfay commented:
I guess the one thing I'd still request is to give names to the TCP_PORT and VIRTUAL_PORT environment variables that are more explicit.  If they can give an idea what they're for it would be cool. They're not just "ports".  Related: [issue about using yaml instead of env vars](https://github.com/drud/ddev/issues/250)

tannerjfco commented:
Would it be clearer if we renamed `VIRTUAL_PORT`->`HTTP_EXPOSE` and `TCP_PORT`->`TCP_EXPOSE` ?

rfay commented:
EXPOSE is a step forward - it implies that we're exposing it on the host.  Maybe you can explain what TCP vs HTTP mean? Now that I read your statement I seem to remember you using nginx for two different types of proxying, HTTP and straight TCP? Because http is tcp of course...

tannerjfco commented:
Users would want to use `HTTP_EXPOSE` if they need to expose a web interface e.g. the admin interface for solr or phpMyAdmin, whereas `TCP_EXPOSE` would be useful for routing non-web traffic, such as mysql (the use case it was introduced for).

rfay commented:
Is it possible that we should be using something like PROXY_STYLE and EXPOSE_PORT instead of bundling the proxy style into the port spec?

beeradb commented:
Using a single PROXY_STYLE assumes only one type of exposed port. IMHO It's probably best to be flexible and allow for both

rfay commented:
No, I was suggesting a PROXY_STYLE=tcp|http|xxx @beeradb - What I'm saying is that the proxy technique is a different piece of information than the port expose information. I could be missing something important though.

tannerjfco commented:
VIRTUAL_PORT and TCP_PORT can accept comma-separated values, so that a container can expose multiple ports. I'm not sure how you would define PROXY_STYLE in a way to account for that without there being some confusion. This approach could also be challenging to handle in template logic for nginx-proxy.

Personally, I'm in favor of the name changes I suggested, along with an update to the documentation to define the two more clearly, and comments for those vars on our docker-compose template.

rfay commented:
Sounds like that's a good path forward then @tannerjfco - and @beeradb likes it!

tannerjfco commented:
This PR is updated per direction in comments above, as is drud/nginx-proxy#10 and #214 for documentation.

rfay commented on pkg/ddevapp/templates.go:
```diff
@@ -40,7 +41,8 @@ services:
       - DEPLOY_NAME=local
       - VIRTUAL_HOST=$DDEV_HOSTNAME
-      - VIRTUAL_PORT=80,{{ .mailhogport }}
+      # HTTP_EXPOSE allows for ports accepting HTTP traffic to be accessible from <site>.ddev.local:<port>
```
Could you please show the colon format example in this lovely comment? Each of the comments could have an example.

rfay commented on pkg/ddevapp/templates.go:
```diff
@@ -12,7 +12,8 @@ services:
     restart: always
     environment:
-      - TCP_PORT=$DDEV_HOSTNAME:{{ .dbport }}
+      # TCP_EXPOSE allows for TCP ports to be accessible from <site>.ddev.local:<port>
```
Please add an example (preferably with colon format) to this lovely comment. Thanks.

rickmanelius commented:
To be clear, it looks like we have a few documentation request from @rfay and addressing the agreed upon PROXY_STYLE item and this will get a green light?

beeradb commented on pkg/plugins/platform/router.go:
```diff
@@ -84,3 +85,56 @@ func StartDockerRouter() {
+		if _, ok := container.Labels["com.ddev.site-name"]; ok {
+			var exposePorts []string
+
+			httpPorts := util.GetContainerEnv("HTTP_EXPOSE", container)
```
Absolutely not recommending we do this right now, but just pointing out that the code here would definitely be simplified by moving to labels for these values.
